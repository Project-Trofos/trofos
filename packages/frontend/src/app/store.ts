import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import trofosApiSlice, { nusmodsApiSlice } from '../api';

const store = configureStore({
  reducer: {
    [trofosApiSlice.reducerPath]: trofosApiSlice.reducer,
    [nusmodsApiSlice.reducerPath]: nusmodsApiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(trofosApiSlice.middleware).concat(nusmodsApiSlice.middleware),
});

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
