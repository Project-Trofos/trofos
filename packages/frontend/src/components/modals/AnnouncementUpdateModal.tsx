import React, { useCallback } from 'react';
import { Form } from 'antd';
import { EditOutlined } from '@ant-design/icons';
import MultistepFormModal from './MultistepModalForm';
import StringFormItem from '../forms/StringFormItem';

type AnnouncementUpdateModalProps = {
  initialTitle: string;
  initialContent: string;
  handleUpdate: (payload: { announcementTitle?: string; announcementContent?: string }) => Promise<void>;
};

/**
 * Modal for updating announcement
 */
export default function AnnouncementUpdateModal(props: AnnouncementUpdateModalProps) {
  const { handleUpdate, initialContent, initialTitle } = props;

  const [form] = Form.useForm();

  const onFinish = useCallback(
    async (values: { announcementTitle?: string; announcementContent?: string }) => {
      const { announcementContent, announcementTitle } = values;
      await handleUpdate({
        announcementTitle,
        announcementContent,
      });
    },
    [handleUpdate],
  );

  return (
    <MultistepFormModal
      title="Update Announcement"
      form={form}
      onSubmit={onFinish}
      buttonChildren={<EditOutlined />}
      buttonSize="small"
      formSteps={[
        <>
          <StringFormItem label="Announcement Title" name="announcementTitle" isRequired initialValue={initialTitle} />
          <StringFormItem
            label="Announcement Content"
            name="announcementContent"
            isRequired
            isTextArea
            initialValue={initialContent}
          />
        </>,
      ]}
    />
  );
}
