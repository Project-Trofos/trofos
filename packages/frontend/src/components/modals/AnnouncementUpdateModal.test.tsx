import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';

import store from '../../app/store';

import AnnouncementUpdateModal from './AnnouncementUpdateModal';

describe('test announcement update modal', () => {
  const initialContent = 'INITIAL_CONTENT';
  const initialTitle = 'INITIAL_TITLE';

  const setup = () => {
    const updateFunc = vi.fn();
    const { baseElement, debug } = render(
      <Provider store={store}>
        <AnnouncementUpdateModal
          initialContent={initialContent}
          initialTitle={initialTitle}
          handleUpdate={updateFunc}
        />
      </Provider>,
    );
    return { baseElement, debug, updateFunc };
  };

  // Antd modal is rendered outside of `root` div and does not get unmounted after closing
  const expectModalInvisible = async (baseElement: HTMLElement) => {
    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const modalWrapper = baseElement.getElementsByClassName('ant-modal-wrap');
    expect(modalWrapper.length).toBe(1);
    await waitFor(() => expect(modalWrapper[0]).toHaveStyle('display: none'));
  };

  it('should render modal with correct fields', () => {
    const { baseElement } = setup();

    // Open modal
    const button = screen.getByRole('button', { name: /open-form/i });
    fireEvent.click(button);

    // Ensure fields are present
    expect(screen.getByLabelText(/announcement title/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/announcement content/i)).toBeInTheDocument();

    // Expect initial values to be shown
    expect(screen.getByDisplayValue(initialTitle)).toBeInTheDocument();
    expect(screen.getByDisplayValue(initialContent)).toBeInTheDocument();

    // Compare with snapshot to ensure structure remains the same
    expect(baseElement).toMatchSnapshot();
  });

  it('should require announcement title', async () => {
    setup();

    const button = screen.getByRole('button', { name: /open-form/i });
    fireEvent.click(button);

    const title = screen.getByLabelText(/announcement title/i);
    fireEvent.change(title, { target: { value: '' } });

    const finishButton = await screen.findByText(/finish/i);
    fireEvent.click(finishButton);

    await screen.findByText('Please input announcement title!');
  });

  it('should require announcement content', async () => {
    setup();

    const button = screen.getByRole('button', { name: /open-form/i });
    fireEvent.click(button);

    const content = screen.getByLabelText(/announcement content/i);
    fireEvent.change(content, { target: { value: '' } });

    const finishButton = screen.getByText(/finish/i);
    fireEvent.click(finishButton);

    await screen.findByText('Please input announcement content!');
  });

  it('should submit correctly if fields are typed in', async () => {
    const { baseElement, updateFunc } = setup();

    const UPDATED_TITLE = 'UPDATED_TITLE';
    const UPDATED_CONTENT = 'UPDATED_CONTENT';

    const button = screen.getByRole('button', { name: /open-form/i });
    fireEvent.click(button);

    const finishButton = await screen.findByText(/finish/i);

    const title = screen.getByLabelText(/announcement title/i);
    fireEvent.change(title, { target: { value: UPDATED_TITLE } });

    const content = screen.getByLabelText(/announcement content/i);
    fireEvent.change(content, { target: { value: UPDATED_CONTENT } });

    fireEvent.click(finishButton);

    // Modal is closed
    await expectModalInvisible(baseElement);

    expect(updateFunc).toBeCalledWith({
      announcementTitle: UPDATED_TITLE,
      announcementContent: UPDATED_CONTENT,
    });
  });
});
