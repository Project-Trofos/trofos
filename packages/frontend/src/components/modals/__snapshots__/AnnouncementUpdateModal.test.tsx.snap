// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test announcement update modal > should render modal with correct fields 1`] = `
<body>
  <div>
    <button
      aria-label="open-form"
      class="ant-btn ant-btn-default ant-btn-color-default ant-btn-variant-outlined ant-btn-sm"
      type="button"
    >
      <span
        aria-label="edit"
        class="anticon anticon-edit"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="edit"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 000-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 009.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div>
    <div
      class="ant-modal-root"
    >
      <div
        class="ant-modal-mask"
      />
      <div
        class="ant-modal-wrap"
        tabindex="-1"
      >
        <div
          aria-labelledby="test-id"
          aria-modal="true"
          class="ant-modal ant-zoom-appear ant-zoom-appear-prepare ant-zoom"
          role="dialog"
          style="width: 520px;"
        >
          <div
            style="outline: none;"
            tabindex="0"
          >
            <div
              class="ant-modal-content"
            >
              <button
                aria-label="Close"
                class="ant-modal-close"
                type="button"
              >
                <span
                  class="ant-modal-close-x"
                >
                  <span
                    aria-label="close"
                    class="anticon anticon-close ant-modal-close-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="close"
                      fill="currentColor"
                      fill-rule="evenodd"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
              <div
                class="ant-modal-header"
              >
                <div
                  class="ant-modal-title"
                  id="test-id"
                >
                  Update Announcement
                </div>
              </div>
              <div
                class="ant-modal-body"
              >
                <form
                  autocomplete="off"
                  class="ant-form ant-form-vertical"
                  id="multi-step-modal-form"
                >
                  <div
                    class="ant-form-item"
                  >
                    <div
                      class="ant-row ant-form-item-row"
                    >
                      <div
                        class="ant-col ant-form-item-label"
                      >
                        <label
                          class="ant-form-item-required"
                          for="multi-step-modal-form_announcementTitle"
                          title="Announcement Title"
                        >
                          Announcement Title
                        </label>
                      </div>
                      <div
                        class="ant-col ant-form-item-control"
                      >
                        <div
                          class="ant-form-item-control-input"
                        >
                          <div
                            class="ant-form-item-control-input-content"
                          >
                            <input
                              aria-required="true"
                              class="ant-input ant-input-outlined"
                              id="multi-step-modal-form_announcementTitle"
                              type="text"
                              value="INITIAL_TITLE"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ant-form-item"
                  >
                    <div
                      class="ant-row ant-form-item-row"
                    >
                      <div
                        class="ant-col ant-form-item-label"
                      >
                        <label
                          class="ant-form-item-required"
                          for="multi-step-modal-form_announcementContent"
                          title="Announcement Content"
                        >
                          Announcement Content
                        </label>
                      </div>
                      <div
                        class="ant-col ant-form-item-control"
                      >
                        <div
                          class="ant-form-item-control-input"
                        >
                          <div
                            class="ant-form-item-control-input-content"
                          >
                            <textarea
                              aria-required="true"
                              class="ant-input ant-input-outlined"
                              id="multi-step-modal-form_announcementContent"
                              rows="4"
                            >
                              INITIAL_CONTENT
                            </textarea>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </form>
              </div>
              <div
                class="ant-modal-footer"
              >
                <button
                  class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                  form="multi-step-modal-form"
                  type="button"
                >
                  <span>
                    Finish
                  </span>
                </button>
              </div>
            </div>
          </div>
          <div
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
